### ISSUE TITLE
too many open files for etcd 2.0 cluster with only a few clients

### ISSUE THREAD

**Issue by yichengq** [coreos/etcd#2728]
https://github.com/coreos/etcd/issues/1959#issuecomment-92685800
https://github.com/coreos/etcd/issues/1959#issuecomment-95045905

```
I'm running a three node etcd cluster which seems to leak file handles (and etcd process memory). When I do lsof for the process I can see ~17000 sockets where lsof can't identify protocol.
```

Could you share the setting of each etcd process, details of etcd clients, and etcd log with me? I have tested it locally and it worked fine to me. No socket boom on proxy or msgdrop on etcd member.

/cc @bahbka @garo

**Comment by garo** [comment]
My setup: Three etcd 2.0.9 in AWS, spread into three AZ us us-east-1. Running in Docker, built from scratch based on public ubuntu:14.04 image and using the 2.0.9 binaries from github: https://github.com/coreos/etcd/releases/download/v2.0.9/etcd-v2.0.9-linux-amd64.tar.gz

Here's startup arguments for one of the instances (copied from docker inspect command):

```
        "Cmd": [
            "./etcd",
            "-heartbeat-interval=500",
            "-election-timeout=3000",
            "-data-dir=/data",
            "-name",
            "etcd-2",
            "-initial-cluster",
            "etcd-1=http://etcd-1:2380,etcd-2=http://etcd-2:2380,etcd-3=http://etcd-3:2380,",
            "-initial-cluster-token",
            "fabric",
            "-initial-advertise-peer-urls",
            "http://etcd-2:2380",
            "-listen-client-urls",
            "http://0.0.0.0:4001,http://0.0.0.0:2379",
            "-listen-peer-urls",
            "http://0.0.0.0:2380",
            "-advertise-client-urls",
            "http://172.16.4.96:2379",
            "-initial-cluster-state",
            "new"
        ],
```

Now the logs: These are from the live cluster.
etcd-2: 232 MiB http://koti.kapsi.fi/garo/nobackup/4b395aed5a19380d0727582b41a109f4df228ac6ca036c0216de60e9db513ca7-json.log
etcd-3: 144 MiB http://koti.kapsi.fi/garo/nobackup/51fa5dd7ac5333ada7b82afe8c6bd0af60dc749373c4d0d8900768cc82037172-json.log

The etcd-2 and etcd-3 has been running for a while. I terminated the old etcd-1 (tried to see if relaunch helps) but I'm currently having some issues with getting it back into the cluster. I'm currently investigating what's wrong with the rejoin.

**Comment by yichengq** [comment]
I still cannot reproduce @bahbka 's case.

@garo I built a etcd 2.0.9 cluster on GCE cluster, and its fd usage is normal(around ten). i let 200 clients connect to each member, and it uses 200 more fds. Then i kill one member, the left two members uses almost the same fd number. I cannot see any of them using 17000~ sockets. Any thoughts?

**Comment by garo** [comment]
@yichengq After much debugging it seems that my issue did originate from my clients which incorrectly left etcd connections open to the cluster, so this might just be an incorrect report from my side. sorry about that.

But how much have we tested etcd with a large number of clients? Do we have statistics on how well for example a three node etcd cluster can work with 100, 1000, 10000, 100000 number of clients? My logs did show a flood of errors which made it hard to debug and it causes a lot of issues, but maybe some of those really were hidden bugs and race conditions inside etcd from the too big amount of clients.

**Comment by yichengq** [comment]
@garo I googled `can't identify protocol` and find out that it happens when the connection is either just inited or closed. I don't know how the etcd process ended with them and wonder whether etcd process could do something to prevent this. Do you know more details about the error? If clients forget to close connections, there should be established connections instead of `can't identify protocol`.

We know that one etcd process could work well with several hundreds of clients. We have the plan for client scalability by using proxy: https://github.com/coreos/etcd/blob/master/Documentation/proxy.md .

**Comment by garo** [comment]
@yichengq I'm not sure what "can't identify protocol" means. It seems that it's visible with lsof but not with netstat -np when I tried to look both to find out more about those sockets.

I like the proxy idea, but we should still have some kind of knowledge how etcd performs under high load. We might identify some errors, leaks or race conditions that way which might be very hard to track down in normal conditions.

**Comment by yichengq** [comment]
@garo Is the client written in go? Could you share it with me?

We are doing it step by step these days. We have functional testing here: https://github.com/coreos/etcd/tree/master/tools/functional-tester to check the correctness under high failure. It is always helpful to do more stress testing.

**Comment by garo** [comment]
I'm using the https://github.com/coreos/go-etcd client in my project https://github.com/garo/orbitcontrol which had an issue that it opened way too much go-etcd client connections and relied that the Go garbage collection would close them in time. I modified my project code structure so that it could open just couple of go-etcd client connections and use them throughout the program execution which seems to fix my issues with it.

**Comment by bahbka** [comment]
How to reproduce my case:
I have 3 virtual servers bahbka-test-1, bahbka-test-2, bahbka-test-3
Using the 2.0.9 binaries from github: https://github.com/coreos/etcd/releases/download/v2.0.9/etcd-v2.0.9-linux-amd64.tar.gz
All servers uses same test start script https://gist.github.com/bahbka/2740f64ff1265d595f7f

bahbka-test-1 log (node): https://gist.github.com/bahbka/cf673dfdb7d39eba1d0c
bahbka-test-2 log (node): https://gist.github.com/bahbka/c8766a37b7f9d574d123
bahbka-test-3 log (proxy): https://gist.github.com/bahbka/98e9872a70c92546c6f1

I think, there is some misconfiguration, but can't understand where.

**Comment by yichengq** [comment]
@garo I see. I have tried to reproduce it with go-etcd. I start 1000 goroutines that keeps NewClient, Get something, and Close. I do see 100+ 'can't identify request' connections at the peak time, but no more than that. Any thoughts on better reproduction?

@bahbka The root problem is that you need to set `-advertise-client-urls` for each etcd member. This is the client urls that proxy will use to do requests.
The default value is http://localhost:2379 and http://localhost:4001, so it connects to itself, and triggers the loop, and failed with too many open files finally. We will improve the behavior.

**Comment by elephantfries** [comment]
All right, the workaround works (2.0.9).  It took me only two days to figure this out :).  Each cluster member not only has to advertise its ip address but it also needs to listen on it and on its localhost as well, like this:

ETCD_LISTEN_CLIENT_URLS="http://10.0.31.78:2379,http://localhost:2379"
ETCD_ADVERTISE_CLIENT_URLS="http://10.0.31.78:2379"

then the proxy can listen on localhost provided it lists the initial cluster, like this:

ETCD_LISTEN_CLIENT_URLS="http://localhost:2379"
ETCD_ADVERTISE_CLIENT_URLS="http://localhost:2379"
ETCD_INITIAL_CLUSTER="mc1=http://10.0.31.103:2380,mc2=http://10.0.31.176:2380,mc3=http://10.0.31.78:2380"

**Comment by yichengq** [comment]
@elephantfries Glad to know you get it. For clarification, this is not the workaround, but the correct way to configure etcd. We expect users to set their listen-client-urls, but it leaves users lots of pains if they don't set it. We have improved both docs and bootstrap process in 2.1.0-alpha.0 for this case.

**Comment by yichengq** [comment]
Based on the discussion, one possibility is etcd client open too many connections at one time, another is the misconfiguration of advertised client urls in etcd member.

The proxy issue will be fixed in #2772
